File: Challenge/May_LeetCoding_Challenge/Week_4:_May_22nd–May_28th/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} preorder
# @return {TreeNode}
def bst_from_preorder(preorder)
  root = TreeNode.new(preorder[0])
  preorder[1..-1].each { |p| add(root, p) }
  root
end

def add(root, p)
  if root.val > p
    root.left ? add(root.left, p) : root.left = TreeNode.new(p)
  else
    root.right ? add(root.right, p) : root.right = TreeNode.new(p)
  end
end

File: Challenge/May_LeetCoding_Challenge/Week_4:_May_22nd–May_28th/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal/solution.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstFromPreorder(preorder []int) *TreeNode {
	root := &TreeNode{Val: preorder[0]}
	for _, val := range preorder[1:] {
		add(root, val)
	}
	return root
}

func add(root *TreeNode, val int) {
	if val < root.Val {
		if root.Left == nil {
			root.Left = &TreeNode{Val: val}
		} else {
			add(root.Left, val)
		}
	} else {
		if root.Right == nil {
			root.Right = &TreeNode{Val: val}
		} else {
			add(root.Right, val)
		}
	}
}